### DISCUSSION TITLE
Keypress ctrl

### DISCUSSION THREAD

**Discussion by Dragonoftheworld** [textadventures/quest#1286] [Creating games with Quest] [answered]
Heyall

So I have this function in js file:
```
var iskeyheld = false;
  $( "body" ).on( "keydown", function(event) {
    if (event.ctrlKey){
      if (iskeyheld == false){
  console.log( "key is held" );
    iskeyheld = true;
     }
   }
 });
 $( "body" ).on( "keyup", function(event) {
   if (event.key === 'Control'){
   console.log( "key is released" );
     if (iskeyheld == true){
     console.log( "key is released" );
       iskeyheld = false;
     }
   }
 });
```
It detects "Key is held" but not both instances of "Key is released".
What am I doing wrong?

**Comment by KVonGit** [comment]
This is not a suggested answer, but it might help. There might already be a "keyup" that you can't override without unbinding it (or something like that -- someone more knowledgeable will probably chime in soon).

I find two places in the code that might be relevant:

https://github.com/textadventures/quest/blob/7a0376151cac36728157d2a49827e783777e8c7e/PlayerController/paper.js#L10116

https://github.com/textadventures/quest/blob/7a0376151cac36728157d2a49827e783777e8c7e/Player/CefKeyboardHandler.vb#L9

**Comment by KVonGit** [comment, marked as answer]
Hello.

No one else saved the day, so I decided to give this another shot.

```js
if (event.key === 'Control'){
```

`event.key` is undefined when I hit CONTROL, but `event.which` is `17`. So, I tried it this way, and it seems to work:

```xml
<!--Saved by Quest 5.8.6836.13983-->
<asl version="580">
  <include ref="English.aslx" />
  <include ref="Core.aslx" />
  <game name="Dragonoftheworld ctrl">
    <gameid>bbc8351c-21a2-4536-825f-467977bc61c2</gameid>
    <version>1.0</version>
    <firstpublished>2024</firstpublished>
    <feature_advancedscripts />
    <inituserinterface type="script">
      JS.eval ("var iskeyheld = false;  $('body').on('keydown', function(event) {    if (event.ctrlKey){      if (iskeyheld == false){  console.log('keydown: key is held');    iskeyheld = true;     }   } });  $('body').on('keyup', function(event) {   if (event.which === 17){   console.log('keyup: key is released');     if (iskeyheld == true){     console.log('keyup: key is released and iskeyheld');       iskeyheld = false;     }   } });")
    </inituserinterface>
  </game>
  <object name="room">
    <isroom />
    <object name="player">
      <inherit name="editor_object" />
      <inherit name="editor_player" />
    </object>
  </object>
</asl>
```

---
NOTE: If I use a JS file instead of using `game.inituserinterface`, none of this works.

**Reply by Dragonoftheworld** [reply]
Hmm that's weird. Good thinking on event.which
```
function checkkeypress(){
var iskeyheld = false;
  $( "body" ).on( "keydown", function(event) {
    if (event.which === 17){
      if (iskeyheld == false){
  console.log( "key is held" );
    iskeyheld = true;
     }
   }
 });
 $( "body" ).on( "keyup", function(event) {
   if (event.which === 17){
   console.log( "key is released" );
     if (iskeyheld == true){
       iskeyheld = false;
     }
   }
 });
}
```
this works in JS file though.🤔

**Reply by KVonGit** [reply]
It might be because Quest 5.8 has such an old browser. The online player might have worked correctly with your original code.

(The browser in Quest 5.9 has been updated. So, we won't have to worry with such things before too long.)
